# Snippy

### generate_snippy_aln_heatmap.py

#### Description

Creates a basic Heatmap from a [Snippy](https://github.com/tseemann/snippy) core.aln or core.full.aln file.
    
Each sample nucleotide position is colored according to the same position in the reference sequence.

Colorscale hex codes:  
Equal to reference: `#ece7f2` ![#ece7f2](https://via.placeholder.com/15/ece7f2/000000?text=+)  
SNP: `#41ab5d` ![#41ab5d](https://via.placeholder.com/15/41ab5d/000000?text=+)  
Zero coverage or deletions: `#0868ac` ![#0868ac](https://via.placeholder.com/15/0868ac/000000?text=+)  
Low coverage, masked region on reference and heterozygous or poor quality genotype: `#252525` ![#252525](https://via.placeholder.com/15/252525/000000?text=+) 

This script requires that 'Biopython' and 'Plotly' be installed.

Note: the output file size will vary according to the number of samples. When using a core.full.aln file, the output
HTML file might be large depending on the number of samples (e.g: a core.full.aln file with the alignment between the
assembly of 1 reference and 7 samples of Streptococcus agalactiae originated a HTML file with 124MB).

#### Usage

Create basic heatmap based on a core.aln or core.full.aln file generated by Snippy:

```
python generate_snippy_aln_heatmap.py -i <snippy_aln_file> -rid <reference_id> -pf <output_html>
```
